File: go/pdfcontext/domain.go
// Package pdfcontext is the pdf bounded context: splitting, PDF-level compression,
// and page/image census. Flat package (idiomatic Go); files split by DDD layer.
// Mirrors python/pdf_toolkit/pdf_context. See ../../DOMAIN.md.
package pdfcontext

import (
	"fmt"

	"pdftoolkit/sharedkernel"
)

// --- Enums -------------------------------------------------------------------

type ImageKind string

const (
	Bitonal   ImageKind = "bitonal"
	Grayscale ImageKind = "grayscale"
	Color     ImageKind = "color"
)

func (k ImageKind) IsContone() bool { return k != Bitonal }

type Codec string

const (
	JPEG    Codec = "jpeg"
	CCITTG4 Codec = "ccitt_g4"
	Flate   Codec = "flate"
	PNG     Codec = "png"
	JBIG2   Codec = "jbig2"
	Raw     Codec = "raw"
)

type ResampleFilter string

const (
	Nearest ResampleFilter = "nearest"
	Lanczos ResampleFilter = "lanczos"
)

// --- Value objects -----------------------------------------------------------

// PageRange is 1-indexed and inclusive.
type PageRange struct {
	Start int
	End   int
}

func NewPageRange(start, end int) (PageRange, error) {
	if start < 1 {
		return PageRange{}, fmt.Errorf("start must be >= 1, got %d", start)
	}
	if end < start {
		return PageRange{}, fmt.Errorf("end (%d) < start (%d)", end, start)
	}
	return PageRange{start, end}, nil
}

func SinglePage(p int) (PageRange, error) { return NewPageRange(p, p) }

func (r PageRange) Pages() []int {
	out := make([]int, 0, r.Len())
	for i := r.Start; i <= r.End; i++ {
		out = append(out, i)
	}
	return out
}

func (r PageRange) Len() int { return r.End - r.Start + 1 }

func (r PageRange) String() string {
	if r.Len() == 1 {
		return fmt.Sprintf("p%d", r.Start)
	}
	return fmt.Sprintf("p%d-%d", r.Start, r.End)
}

type SplitSpec struct {
	Name  string
	Range PageRange
}

// EmbeddedImage is one raster image on a page.
type EmbeddedImage struct {
	Xref             int
	PageIndex        int
	Width            int
	Height           int
	Kind             ImageKind
	Codec            Codec
	EffectiveDpi     sharedkernel.EffectiveDpi
	RenderedAreaSqIn float64
}

// Resample enforces hard rule #2: bitonal must not be interpolated into gray.
func (e EmbeddedImage) Resample() ResampleFilter {
	if e.Kind == Bitonal {
		return Nearest
	}
	return Lanczos
}

// DocumentCensus is the output of inspect: what's inside a PDF.
type DocumentCensus struct {
	File          sharedkernel.MediaFile
	PageCount     int
	Images        []EmbeddedImage
	NonImageBytes int64
}

func (c DocumentCensus) TotalRenderedArea() float64 {
	var total float64
	for _, img := range c.Images {
		total += img.RenderedAreaSqIn
	}
	if total == 0 {
		return 1.0
	}
	return total
}

func (c DocumentCensus) ImageBytesFraction() float64 {
	total := c.File.SizeBytes()
	if total == 0 {
		return 0
	}
	return float64(total-c.NonImageBytes) / float64(total)
}

type CompressionTarget struct {
	Budget       sharedkernel.ByteBudget
	QualityFloor sharedkernel.QualityFloor
	DpiCap       int
	DpiRange     [2]int
	QualityRange [2]int
}

func DefaultTarget(budget sharedkernel.ByteBudget) CompressionTarget {
	return CompressionTarget{
		Budget:       budget,
		QualityFloor: sharedkernel.DefaultQualityFloor(),
		DpiCap:       200,
		DpiRange:     [2]int{72, 300},
		QualityRange: [2]int{20, 95},
	}
}

type CompressionRecipe struct {
	Xref     int
	DpiTarget int
	Quality  int
	Codec    Codec
	Resample ResampleFilter
}

type CompressionResult struct {
	Output      sharedkernel.MediaFile
	Engine      string
	BeforeBytes int64
	AfterBytes  int64
	DpiUsed     int
	QualityUsed int
	Score       *sharedkernel.PerceptualScore
	ElapsedMs   int64
	Escalated   bool
}

func (r CompressionResult) Ratio() float64 {
	if r.BeforeBytes == 0 {
		return 1.0
	}
	return float64(r.AfterBytes) / float64(r.BeforeBytes)
}

func (r CompressionResult) SavedPct() float64 { return 100 * (1 - r.Ratio()) }

// --- Domain events -----------------------------------------------------------

type CompressionAttempted struct {
	File       string
	Engine     string
	AfterBytes int64
}

type BudgetMissed struct {
	File       string
	Engine     string
	AfterBytes int64
	Ceiling    int64
}

type QualityFloorViolated struct {
	File  string
	Score sharedkernel.PerceptualScore
	Floor sharedkernel.QualityFloor
}

File: rust/src/pdf_context/domain.rs
//! pdf context — domain layer. Mirrors python/pdf_toolkit/pdf_context/domain.

use crate::shared_kernel::{ByteBudget, EffectiveDpi, MediaFile, PerceptualScore, QualityFloor};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ImageKind {
    Bitonal,
    Grayscale,
    Color,
}

impl ImageKind {
    pub fn is_contone(&self) -> bool {
        !matches!(self, ImageKind::Bitonal)
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Codec {
    Jpeg,
    CcittG4,
    Flate,
    Png,
    Jbig2,
    Raw,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ResampleFilter {
    Nearest,
    Lanczos,
}

/// 1-indexed, inclusive.
#[derive(Clone, Copy, Debug)]
pub struct PageRange {
    pub start: u32,
    pub end: u32,
}

impl PageRange {
    pub fn new(start: u32, end: u32) -> Result<Self, String> {
        if start < 1 {
            return Err(format!("start must be >= 1, got {start}"));
        }
        if end < start {
            return Err(format!("end ({end}) < start ({start})"));
        }
        Ok(Self { start, end })
    }
    pub fn single(p: u32) -> Result<Self, String> {
        Self::new(p, p)
    }
    pub fn pages(&self) -> Vec<u32> {
        (self.start..=self.end).collect()
    }
    pub fn len(&self) -> u32 {
        self.end - self.start + 1
    }
    pub fn is_empty(&self) -> bool {
        false
    }
    pub fn label(&self) -> String {
        if self.len() == 1 {
            format!("p{}", self.start)
        } else {
            format!("p{}-{}", self.start, self.end)
        }
    }
}

#[derive(Clone, Debug)]
pub struct SplitSpec {
    pub name: String,
    pub range: PageRange,
}

/// One raster image on a page.
#[derive(Clone, Copy, Debug)]
pub struct EmbeddedImage {
    pub xref: u32,
    pub page_index: u32,
    pub width: u32,
    pub height: u32,
    pub kind: ImageKind,
    pub codec: Codec,
    pub effective_dpi: EffectiveDpi,
    pub rendered_area_sqin: f64,
}

impl EmbeddedImage {
    /// Hard rule #2: bitonal must not be interpolated into gray.
    pub fn resample(&self) -> ResampleFilter {
        if matches!(self.kind, ImageKind::Bitonal) {
            ResampleFilter::Nearest
        } else {
            ResampleFilter::Lanczos
        }
    }
}

/// Output of inspect: what's inside a PDF.
#[derive(Clone, Debug)]
pub struct DocumentCensus {
    pub file: MediaFile,
    pub page_count: u32,
    pub images: Vec<EmbeddedImage>,
    pub non_image_bytes: u64,
}

impl DocumentCensus {
    pub fn total_rendered_area(&self) -> f64 {
        let t: f64 = self.images.iter().map(|i| i.rendered_area_sqin).sum();
        if t == 0.0 {
            1.0
        } else {
            t
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct CompressionTarget {
    pub budget: ByteBudget,
    pub quality_floor: QualityFloor,
    pub dpi_cap: u32,
    pub dpi_range: (u32, u32),
    pub quality_range: (u32, u32),
}

impl CompressionTarget {
    pub fn with_budget(budget: ByteBudget) -> Self {
        Self {
            budget,
            quality_floor: QualityFloor::default(),
            dpi_cap: 200,
            dpi_range: (72, 300),
            quality_range: (20, 95),
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct CompressionRecipe {
    pub xref: u32,
    pub dpi_target: u32,
    pub quality: u32,
    pub codec: Codec,
    pub resample: ResampleFilter,
}

#[derive(Clone, Debug)]
pub struct CompressionResult {
    pub output: MediaFile,
    pub engine: String,
    pub before_bytes: u64,
    pub after_bytes: u64,
    pub dpi_used: u32,
    pub quality_used: u32,
    pub score: Option<PerceptualScore>,
    pub elapsed_ms: u128,
    pub escalated: bool,
}

impl CompressionResult {
    pub fn ratio(&self) -> f64 {
        if self.before_bytes == 0 {
            1.0
        } else {
            self.after_bytes as f64 / self.before_bytes as f64
        }
    }
    pub fn saved_pct(&self) -> f64 {
        100.0 * (1.0 - self.ratio())
    }
}
